lib/utils/utf8.c

```diff
@@ -79,3 +84,34 @@ char *utf8_lcpy(char *dst, const char *src, size_t n)
 
 	return dst;
 }
+
+bool utf8_is_valid(const char *str, size_t maxlen)
+{
+	size_t i = 0, nbyte = 0;
+	size_t len = strnlen(str, maxlen);
```

Comment:
Given the challenges with `strnlen`, could we just do the following? Or would that be considered insecure? 

```suggestion
	size_t len = MIN(strlen(str), maxlen);
```

Reply:
I would consider that insecure. Why not do the check when looping?

Reply:
> Why not do the check when looping?

Can you elaborate?

Reply:
```c
	while (i < len) {
		if (str[i] == '\0') {
			break;
		}

		if (str[i] <= ASCII_CHAR && str[i] > '\0') {
			i++;
			continue;
		}

		// ...
	}
```

Reply:
I'm sorry to tell you, that empty characters also belong to utf8 characters. If you do this, the check will be cut off.

Reply:
The check would already be cut off, as `strlen` is simply looking for a `\0` byte, no?

Reply:
> UTF-8 is an encoding that is used to represent multibyte character sets in a way that is backward-compatible with single-byte character sets. Another advantage of UTF-8 is that it ensures there are no NULL bytes in the data, with the exception of an actual NULL byte.

From https://www.oreilly.com/library/view/secure-programming-cookbook/0596003943/ch03s12.html#:~:text=Discussion,of%20an%20actual%20NULL%20byte.

Reply:
Agreed. 

[0x30, 0x31, 0x00, 0x30, 0xff] should be considered a valid UTF8 string as only the first 2 characters (0x30, 0x31) and the NULL terminator should be considered. 

For example
```
char str[100];
str[0] = 0x30; /* or '0' */
str[1] = 0x31; /* or '1' */ 
str[2] = 0x00; /* or '\0' */
```

is a valid UTF8 string, even if octets [3..99] may be invalid, since the string stops at the NULL terminator. 

I like the suggestion from @pdgendt as it also optimizes the check by avoiding the call to `strlen`.

Reply:
Yes. So we don't need to loop the empty character into looping.